### PR TITLE
Env: correct multisite support

### PR DESCRIPTION
**Pull request by noahtallen**
## Description
So it turns out that wp-env isn't really doing multisite tests right now. After some debugging by @TimothyBJacobs, we discovered that our wp-config.php requires wp-settings.php _before_ MULTISITE is defined to true. This means that wp-phpunit can't redefine that constant to true. Normally, wp-phpunit would define the constant to true if there is an existing MULTISITE env variable and _then_ load the wp-settings.php file.

Basically, wp-settings.php is being loaded twice: once by us and once by wp-phpunit. We can solve the multisite issue (and other issues, I'm sure) by not loading it ourselves when running wp-phpunit. I've done this by creating a copy of wp-config.php which does not load wp-settings.php when wordpress is initialized by wp-env.

One nice bonus is that this is also the cause of our "redeclared constant" errors. Before this PR, you'd see 3-4 of those when running phpunit, and now there are none. This makes a lot of sense if these files have been loaded twice!

Unfortunately, after fixing this, I am getting the following error running multisite tests. This error does not appear for single site test runs.
```sh
$ npm run test-unit-php-multisite

> gutenberg@8.2.0-rc.1 test-unit-php-multisite /Users/noah.allen/source/gutenberg
> wp-env run phpunit 'WP_MULTISITE=1 phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist'

✖ Error while running docker-compose command.
Installing...
Installing network...
Running as multisite...

wp_die called
Message : <h1>Error establishing a database connection</h1>
Title : Database Error

wp_die called
Message : <h1>Error establishing a database connection</h1><p>If your site does not display, please contact the owner of this network. If you are the owner of this network please check that MySQL is running properly and all tables are error free.</p><p><strong>Could not find site <code>localhost:8889</code>.</strong> Searched for table <code>wp_blogs</code> in database <code>tests-wordpress</code>. Is that right?</p><p><strong>What do I do now?</strong> Read the <a href="https://wordpress.org/support/article/debugging-a-wordpress-network/" target="_blank">bug report</a> page. Some of the guidelines there may help you figure out what went wrong. If you&#8217;re still stuck with this message, then check that your database contains the following tables:</p><ul><li>wp_users</li><li>wp_usermeta</li><li>wp_blogs</li><li>wp_blogmeta</li><li>wp_signups</li><li>wp_site</li><li>wp_sitemeta</li><li>wp_registration_log</li></ul>
Title : Error establishing a database connection
Args:
	 response : 500
```
## How has this been tested?
1. Made sure that wp-env start works after wp-env destroy, and that it works running wp-env start multiple times in a row.
2. phpunit tests should pass locally and in CI.

## Types of changes
bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by epiqueras** [comment]
Have you tried just hardcoding `MULTISITE` in the original `wp-config.php` instead of creating a copy that doesn't load settings?

**Comment by noahtallen** [comment]
> Have you tried just hardcoding MULTISITE in the original wp-config.php 

Yeah, I also think that would solve the original problem, but we probably wouldn’t want it turned on all the time. The reason we went with this route is that there could be other things that should be configured before loading settings, so it might be hard to pick all of those out by hand to define early. Additionally, loading settings twice causes all of the redeclared constant warnings you see running phpunit in wp-env. We would still probably not want to load the file twice

**Comment by noahtallen** [comment]
Note that the multisite tests are failing in this PR even though travis is green. This is because the phpunit command exits with 0 even though the database connection failed.

**Comment by noahtallen** [comment]
After some debugging it looks like wp_die is called at `require_once ABSPATH . '/wp-settings.php';` in wp-phpunit's bootstrap.php.

**Comment by noisysocks** [comment]
Could this be fixed in `wp-phpunit`? I'm probably misunderstanding something, but my first impression is that it's strange that one has to deviate from the canonical<sup>1</sup> way to write a `wp-config.php` file in order to get tests to work.

_<sup>1</sup> In that it's what `wp-config-sample.php` tells you to do._

**Comment by noahtallen** [comment]
@noisysocks I think this is a shortcoming of wp-phpunit, but it's an explicit one. It has us set the path to our own config file, and then it loads our config file in its own config file

**Comment by noahtallen** [comment]
see the "why" section here: https://github.com/wp-phpunit/docs#overview

> WP PHPUnit solves this by including this file in the package source which acts as a kind of proxy to your "real" tests config file, which can be wherever you wish.

**Comment by noahtallen** [comment]
> Could this be fixed in wp-phpunit?

I'm also conflicted here because so much about phpunit is tied to the environment (wp-env), but wp-env's primary goal isn't to support 3rd party deps like phpunit. To make it straightforward to run phpunit, we have to modify wp-env, but that means accepting phpunit under wp-env's umbrella even though it's currently a composer dependency of the Gutenberg plugin.

**Comment by TimothyBJacobs** [comment]
> my first impression is that it's strange that one has to deviate from the canonical1 way to write a wp-config.php file in order to get tests to work.

WP also ships with a separate `wp-tests-config-sample.php` that is somewhat different from the plain `wp-config.php` format.

> wp-env's primary goal isn't to support 3rd party deps like phpunit.

I think `wp-env` wouldn't have to work so hard if WordPress played with phpunit in a more standard way like other projects do. But because it doesn't, I think it isn't so much `wp-env` supporting the 3rd party tool phpunit, but rather supporting the first-party tool of the WordPress unit test suite if that makes sense.

**Comment by noisysocks** [comment]
> see the "why" section here: https://github.com/wp-phpunit/docs#overview

Thanks! That's helpful 🙂

> To make it straightforward to run phpunit, we have to modify wp-env, but that means accepting phpunit under wp-env's umbrella even though it's currently a composer dependency of the Gutenberg plugin.

Yeah... are we conflating Gutenberg and `wp-env`? Gutenberg depends on `wp-phpunit` and `wp-env`. `wp-env` depends on `phpunit`. `wp-env` should **not** depend on `wp-phpunit`. Gutenberg should invoke tests by running `WP_PHPUNIT__TESTS_CONFIG=phpunit/wp-tests-config.php wp-env run phpunit`. `phpunit/wp-tests-config.php` can be written with the assumption that the test environment is using the default `wp-env` port etc.

**Comment by noisysocks** [comment]
> I think wp-env wouldn't have to work so hard if WordPress played with phpunit in a more standard way like other projects do.

It's not really relevant to this PR, but I'm curious what changes to WordPress would make it play nicer with `phpunit`? We have the power to change Core 🙂

**Comment by noahtallen** [comment]
> `wp-env` should not depend on `wp-phpunit`

why not? As we've seen already, setting up phpunit is a pain, and it's good for us to support unit-tested code in the WordPress ecosystem. It'd be great to make it less of a pain, IMO. I just wonder if wp-env _should_ be trying to absorb this complexity since the phpunit setup is so closely tied to the other configuration code in wp-env

**Comment by TimothyBJacobs** [comment]
Pushed a commit to set the additional test constants that multisite needs.

> It's not really relevant to this PR, but I'm curious what changes to WordPress would make it play nicer with phpunit?

I think a big thing is that `wp-config.php` has [side effects](https://core.trac.wordpres.org/ticket/9444) which means we need to go to all this trouble. Also that all these constants need to be set which sort of necessitates a separate tests config file.

**Comment by noahtallen** [comment]
Hm, still seeing the database connection here: https://travis-ci.com/github/WordPress/gutenberg/jobs/339864292

**Comment by TimothyBJacobs** [comment]
Yeah it's strange. Before that commit, the tests blew up entirely for me, and after that commit they work. But that seemed to have no effect in travis?

File: .travis.yml [hunk 1]
```diff
@@ -128,13 +128,11 @@ jobs:
         - npm run test-unit:native -- --ci --maxWorkers=2 --cacheDirectory="$HOME/.jest-cache"
 
     - name: PHP unit tests
-      env: WP_ENV_TESTS_PORT=8887 WP_ENV_PORT=8886 # TODO: Remove tests port when scripts/env is replaced with wp-env.
       script:
         - chmod -R 767 ./build
         - npm run test-php && npm run test-unit-php-multisite
 
     - name: PHP unit tests (PHP 5.6)
-      env: LOCAL_PHP=5.6-fpm WP_ENV_TESTS_PORT=8887 WP_ENV_PORT=8886 # TODO: Remove tests port when scripts/env is replaced with wp-env.
```
Comment (noahtallen) [inline .travis.yml:137]: we still need the LOCAL_PHP env var here to set the php version

**Comment by epiqueras** [comment]
>Yeah, I also think that would solve the original problem, but we probably wouldn’t want it turned on all the time.

I was just wondering if that approach at least gets the multisite tests to pass.

**Comment by noahtallen** [comment]
> if that approach at least gets the multisite tests to pass.

_Theoretically_, it shouldn't change behavior, since wp-phpunit does define the constant if it exists as an environment variable. We've confirmed that behavior works, but it's worth a shot

**Comment by noahtallen** [comment]
I think a good first step here is to make sure the multisite test fails if:
1. it's not actually multisite
2. the database connection fails

If either of these fail, we currently don't know about it without looking at the logs since travis passes.

File: .travis.yml (hunk 1 above)
Comment (TimothyBJacobs) [reply .travis.yml:137]: Ack my mistake.

**Comment by TimothyBJacobs** [comment]
AFAICT, we are at the point that the multisite tests are actually passing and working on multisite. The issue is the db error output which I think is happening because it isn't able to find the right current site.

File: .travis.yml (hunk 1 above)
Comment (noahtallen) [reply .travis.yml:137]: (btw just pushed a fix)

**Comment by noahtallen** [comment]
Another oddity is that if I add a wp_die call in `includes/install.php`, it does fail the test run for me locally. It also fails if wp_die is added to a random test. So I'm not sure where this one is happening that it avoids the normal exception behavior

**Comment by noahtallen** [comment]
the current failure is a flakey e2e test, not php. Don't let that fool you into thinking multisite php tests are working because they aren't :trollface:

**Comment by noahtallen** [comment]
Ok, so the test correctly fails now if installation did not work. This also gives us a stacktrace:

```sh
Fatal error: Uncaught Exception: WordPress died. in /var/www/html/wp-content/plugins/gutenberg/phpunit/bootstrap.php:58
Stack trace:
#0 /var/www/html/wp-includes/class-wp-hook.php(287): fail_if_died('_default_wp_die...')
#1 /var/www/html/wp-includes/plugin.php(206): WP_Hook->apply_filters('_default_wp_die...', Array)
#2 /var/www/html/wp-includes/functions.php(3360): apply_filters('wp_die_handler', '_default_wp_die...')
#3 /var/www/html/wp-includes/functions.php(4666): wp_die('<h1>Error estab...', 'Database Error')
#4 /var/www/html/wp-includes/ms-load.php(468): dead_db()
#5 /var/www/html/wp-includes/ms-settings.php(77): ms_not_installed('localhost:8889', '')
#6 /var/www/html/wp-settings.php(141): require('/var/www/html/w...')
#7 /var/www/html/wp-content/plugins/gutenberg/vendor/wp-phpunit/wp-phpunit/includes/bootstrap.php(141): require_once('/var/www/html/w...')
#8 /var/www/html/wp-content/plugins/gutenberg/phpunit/bootstrap.php(63): require('/var/www/html/w...')
#9 phar:///usr/local/bin/phpunit/phpunit/Util/FileLoader.php in /var/www/html/wp-content/plugins/gutenberg/phpunit/bootstrap.php on line 58
``

File: .travis.yml [hunk 2]
```diff
@@ -54,6 +54,7 @@ install:
   - |
     if [[ "$INSTALL_WORDPRESS" = "true" ]]; then
       echo '{ "config": { "SCRIPT_DEBUG": false, "WP_PHP_BINARY": "php" } }' > .wp-env.override.json
+      chmod -R 767 ./ # TODO: Possibly integrate in wp-env
```
Comment (noahtallen) [inline .travis.yml:57]: should probably make this go back to doing ./build. doing all of cwd took a long time for me locally. (plus we should figure out why we need this)

**Comment by jeremyfelt** [comment]
I'm not familiar enough with wp-env to know if this is related, but it could be: Multisite does not support custom port numbers. See https://core.trac.wordpress.org/ticket/21077 and https://core.trac.wordpress.org/ticket/15936 for background.

If the address being used to start up the tests is `localhost:8889`, then there's a good chance the database error is not really a _database_ error, but is just WordPress not finding the site that was requested or bailing for some other reason.

I do know that in November I was not able to get the new core development environment introduced in [Trac#47767](https://core.trac.wordpress.org/ticket/47767) working in multisite. I'm guessing it and wp-env are either the same or very related, but I haven't paid close enough attention. 🙈 

Hope that's helpful! Happy to try to point in other directions as well. :)

**Comment by spacedmonkey** [comment]
This [patch](https://core.trac.wordpress.org/attachment/ticket/21077/15936.5.diff) in [#21077](https://core.trac.wordpress.org/ticket/21077) is ready for merge / core. It doesn't fix the issue, but adds a filter that allow the issue to be fixed via a filter. 

Anyone with commit, can commit. Once that is in core, I will support to get this working in this PR.

**Comment by noahtallen** [comment]
> Multisite does not support custom port numbers

that definitely sounds related. I wonder why it would still work locally for us, though 🤔 I'll try a solution for this, though!

**Comment by noahtallen** [comment]
I tried locally by changing the tests port to `443`, but it still seems to be failing, though it now fails with ` ms_not_installed('localhost', '')` (i.e. excluding the port number from localhost)

**Comment by TimothyBJacobs** [comment]
I fixed my debugging code to properly dump the blogs table right before we die. It [looks like](https://travis-ci.com/github/WordPress/gutenberg/jobs/343006076#L850) the blogs table isn't even being populated. I think that means that `populate_network` is failing somewhere?

I opened a ticket for adjusting this in core, https://core.trac.wordpress.org/ticket/50251.

The `install.php` file is brought in via composer which makes adjusting it particularly difficult. I've tried including a modified version and copying it, but this, of course, fails on travis due to permission issues.

**Comment by TimothyBJacobs** [comment]
Was finally able to get an error message to appear

```
Error populating network: You must provide a valid email address.
```

We did add this constant though in `packages/env/lib/config.js`.

**Comment by noahtallen** [comment]
cc @epiqueras @noisysocks we have the tests passing now. Just had to copy the test config values to the override file! should be good for a review 👍

**Review by epiqueras — APPROVED** [review]

**Comment by epiqueras** [comment]
Great work here!

File: .travis.yml (hunk 2 above)
Comment (youknowriad) [reply .travis.yml:57]: Do you know why this "chmod" is needed?

Comment (noahtallen) [reply .travis.yml:57]: Yes, some of the tests needed filesystem write access, which was not working until we added this. Did you end up needing it during the switch to GitHub workflows?

**Pull request merged**